mlir/lib/Dialect/MIOpen/CoordTransformBuilder.cpp

```diff
@@ -523,6 +553,55 @@ void BottomUpCTBuilder::addDim(StringRef name, uint32_t dim, int64_t size) {
   addTransform(TransformType::AddDim, {size}, {}, {}, {name}, {dim});
 }
 
+void BottomUpCTBuilder::expand(ArrayRef<uint32_t> dims, ArrayRef<int64_t> sizes) {
+  SmallVector<StringRef, 8> ptNames;
+  SmallVector<uint32_t, 8> ptDims;
+  uint32_t dim = 0;
+  for (uint32_t i = 0; i < nStartDims(); ++i) {
+    ptNames.push_back(nStartNames()[i]);
+    while (std::find(dims.begin(), dims.end(), dim) != dims.end())
+      dim++;
+    ptDims.push_back(dim++);
+  }
+  passThrough(ptNames, ptDims, ptNames);
+
+  for (auto tuple : llvm::zip(dims, sizes)) {
+    auto dim = std::get<0>(tuple);
+    auto size = std::get<1>(tuple);
+    SmallString<8> name;
+    ("exp" + Twine(dim)).toVector(name);
+
+    addDim(name, dim, size);
+  }
+}
+
+void BottomUpCTBuilder::broadcast(ArrayRef<uint32_t> bcastDims,
```

Comment:
No, I don't agree with this signature. It should be (with some specialized versions, versions that don't mention names, etc.) `BottomUpCTBuilder::broadcast(ArrayRef<StringRef> endNames, ArrayRef<uint32_t> endDims, ArrayRef<StringRef> startNames, ArrayRef<int64_t> upperSizes)`

and also `TopDownCTBuilder::collapse(ArrayRef<StringRef> endNames, ArrayRef<uint32_t> endDims, ArrayRef<StringRef> startNames, ArrayRef<int64_t> lowerSizes)` which also produces a broadcast

Reply:
Alternatively, allowing for fun times like, `[broadcast{1} ["d"] at [0] -> ["d"] at [0]] bounds [4] -> [2]` being the transform [a, a, a, a] <- [a, b], the `broadcast()` method could be most general by taking both the upper and lower sizes, but that's a stretch and I'm not entirely sold on it either.

Reply:
Agreed, hence why I was hesitant about the general case. Just passing in the upper dim sizes might be sufficient, if we are willing to make the assumption that we always broadcast the lower dimension in its entirity - we could then just copy those lower sizes to the parameters. (My instinct towards generality says we should provide a way to override that default, but still).

I do, at the very least, want broadcast() to allow for the lower and upper dimensions to be in different places.